### ISSUE TITLE
What's the best way to run tests in parallel using same persistent context

### ISSUE THREAD

**Issue by ankit-scio** [microsoft/playwright#7146]
We use bazel for all our build and test. I have setup a bunch of bazel test targets which all use playwright to test our product's UI. This is the flow I want to achieve:

1. Run one bazel test which would login to the product
2. All other bazel tests run in parallel using the same logged in state from the first test

Things I have tried so far:
1. Use persistent context: 1st test launches a persistent context pointing to a user directory and logs in. All other tests after that launch persistent context pointing to the same user directory in parallel. This doesn't work as launchPersistentContext works only for the first test in the parallel set and fails for the rest.
2. Store local and session storage: Have tried the approach mentioned in https://playwright.dev/docs/auth to save and retrieve the local and session storage for my app, but looks like this doesn't avoid the first login page in my app where the user email is asked. I still have to enter the email and then it takes me to the home page. So may have to also look at storing indexedDB here, which I don't currently know how to.

These bazel tests create a separate sandboxed build files for each test, so not able to use the same browser context instance and have tests just create a new page on it instead of having to launch a browser context per test.

**Comment by pavelfeldman** [comment]
Did you look at https://playwright.dev/docs/auth#reuse-authentication-in-playwright-test specifically? It sounds exactly like the scenario you are describing.

**Comment by ankit-scio** [comment]
Yeah as mentioned in point 2 of my message, have already tried the local and session storage save and load way to bypass auth. But for my application, this doesn't seem to work. My eng team has suggested to also save and load indexedDB, so trying that out.

The persistent context on the other hand works perfectly, but doesn't allow me to run tests in parallel using the same user data dir.

**Comment by ankit-scio** [comment]
For me this works fine with persistent context https://playwright.dev/docs/auth/#lifecycle Only if this could work with tests creating browser contexts pointing to the same user data dir in parallel

**Comment by ankit-scio** [comment]
In playwright tests do we have support for a flow like this

1. In global setup create a persistent browser context and log in to the product
2. All tests either load a page on the same browser context instead of creating sandboxed browsers or are able to use the same persistent context and are run in parallel

**Comment by pavelfeldman** [comment]
You can make it work manually, but that isn't something we would want our users to do. Can you tell more about your use case? Is it for testing? Why does regular context not work for you?

**Comment by ankit-scio** [comment]
Yes it is for e2e testing.

Workflow:
1. Launch a browser context with cookie/local storage already loaded
2. Goto homepage of our app
3. Redirects to login page still - This is because the product still wants to know which backend to hit which is dependant on the email domain that we enter on this page.
4. Enter email and submit - Without the cookie/local storage the redirect to SSO login would happen. But because we already have that set, we directly go to the app home page

The problem is that our product has issues on step 4 because of the missing redirect to the SSO. And we end up intermittently on an eternal loading screen with a spinner.

To avoid this, I checked with the product team on how we can avoid the login page on step 3 completely. And for that will need to also have the indexedDB loaded on the context, which I am still trying to figure out how to properly save and restore. Also, using a persistent context as I described in the main description works but I lose running tests in parallel as our bazel test launches each spec in a completely sandboxed env.

Do you know of a way to save and restore indexedDB?

**Comment by aslushnikov** [comment]
> Do you know of a way to save and restore indexedDB?

@ankit-scio Both Chromium and WebKit support the [`indexedDB.databases()`](https://developer.mozilla.org/en-US/docs/Web/API/IDBFactory/databases) call to retrieve all databases for the current origin. Once you have names of databases, you can store/export indexedDB using the snippet here:

https://gist.github.com/loilo/ed43739361ec718129a15ae5d531095b

**Comment by ankit-scio** [comment]
> > Do you know of a way to save and restore indexedDB?
> 
> @ankit-scio Both Chromium and WebKit support the [`indexedDB.databases()`](https://developer.mozilla.org/en-US/docs/Web/API/IDBFactory/databases) call to retrieve all databases for the current origin. Once you have names of databases, you can store/export indexedDB using the snippet here:
> 
> https://gist.github.com/loilo/ed43739361ec718129a15ae5d531095b

Thanks for the pointer. Will check it out.

**Comment by dgozman** [comment]
@ankit-scio Did that indexedDB snippet work for you?

**Comment by ankit-scio** [comment]
@dgozman Thanks for following up. No haven't tried this yet. Will do by the end of this month and update. Thanks

**Comment by aslushnikov** [comment]
I'll close this for now; please feel free to file a new issue with an update if things don't work for you.
